### PR TITLE
reduce variance in AutoDAIS

### PR DESCRIPTION
**Pull request by martinjankowiak**
@fehiepsi can you please check the algebra on this? it's very simple

basically we get a recursion `<v_{k-1}> = gamma ( <v_{k-2}> + g_{k-1} ) `where `g_k` is the gradient multiplied by eta.

also the kinetic energy difference is proportional to `dot(g_k * M^{-1}, 2 v_{k-1} + g_k)`

i checked that we get the same elbo (and lower variance) for `infer/test_autoguide.py::test_beta_bernoulli`

also compared this PR against master for a larger logistic regression dataset and this PR led to a better elbo after 25k and 80k iterations (i.e. elbo increasing more quickly per iteration). though this PR is slightly slower because of the extra carry/math

### CODE REVIEW

File: numpyro/infer/autoguide.py [hunk 1]
```diff
@@ -825,19 +828,23 @@ def scan_body(carry, eps_beta):
...
             z_half = z_prev + v_prev * eta * inv_mass_matrix
             q_grad = (1.0 - beta) * grad(base_z_dist.log_prob)(z_half)
             p_grad = beta * grad(log_density)(z_half)
-            v_hat = v_prev + eta * (q_grad + p_grad)
+            eta_grad = eta * (q_grad + p_grad)
+            v_hat = v_prev + eta_grad
             z = z_half + v_hat * eta * inv_mass_matrix
             v = gamma * v_hat + jnp.sqrt(1 - gamma ** 2) * eps
-            delta_ke = momentum_dist.log_prob(v_prev) - momentum_dist.log_prob(v_hat)
+            delta_ke = jnp.dot(eta_grad * inv_mass_matrix, 2.0 * v_prev_mean + eta_grad)
```
[the remark this answers is not in the document]
Comment (martinjankowiak) [reply numpyro/infer/autoguide.py:839]: hmm, yes, you're right, whoops. i guess what i would really need to do is to treat this as a control variate, which would involve doing a first order taylor expansion of `g_k` around `z_{k-1}`. but i guess this probably (?) wouldn't be worth the extra computational cost. thanks for checking!
